packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -356,31 +351,134 @@ export class ScheduleManager {
-        if (this.batchClientId !== message.clientId) {
-            // As a back stop for any bugs marking the end of a batch - if the client ID flipped, we
-            // consider the previous batch over.
-            if (this.batchClientId) {
-                this.emitter.emit("batchEnd", "Did not receive real batchEnd message", undefined);
-                this.deltaScheduler.batchEnd();
+    private setPaused(localPaused: boolean) {
+        // Return early if no change in value
+        if (this.localPaused === localPaused) {
+            return;
+        }
+
+        this.localPaused = localPaused;
+        if (localPaused) {
+            // eslint-disable-next-line @typescript-eslint/no-floating-promises
+            this.deltaManager.inbound.pause();
+        } else {
+            this.deltaManager.inbound.resume();
+        }
+    }
+
+    /**
+     * Called for each incoming op (i.e. inbound "push" notification)
+     */
+    private trackPending(message: ISequencedDocumentMessage) {
+        assert(this.deltaManager.inbound.length !== 0, "we have something in the queue that generates this event");
+
+        const metadata = message.metadata as IRuntimeMessageMetadata;
+
+        // Protocol messages are never part of a runtime batch of messages
+        if (!isRuntimeMessage(message)) {
+            // Protocol messages should never show up in the middle of the batch!
+            assert(this.currentBatchClientId === undefined, "System message in the middle of batch!");
+
+            this.pauseSequenceNumber = undefined;
+            this.currentBatchClientId = undefined;
+            this.setPaused(false);
+            return;
+        }
+
+        const batchMetadata = metadata?.batch;
+
+        // If the client ID changes then we can move the pause point. If it stayed the same then we need to check.
+        // If batchMetadata is not undefined then if it's true we've begun a new batch - if false we've ended
+        // the previous one
+        if (this.currentBatchClientId !== undefined || batchMetadata === false) {
+            assert(this.currentBatchClientId === message.clientId,
+                "Batch not closed, yet message from another client!");
+        }
+        if (batchMetadata) {
+            assert(this.currentBatchClientId === undefined, "there can't be active batch");
+            this.pauseSequenceNumber = message.sequenceNumber;
+            this.currentBatchClientId = message.clientId;
+            this.setPaused(true);
+        } else if (batchMetadata === false) {
+            assert(this.localPaused, "active batch, should be paused");
+            this.pauseSequenceNumber = undefined;
+            this.currentBatchClientId = undefined;
+            this.setPaused(false);
+        } else {
+            this.setPaused(this.pauseSequenceNumber !== undefined);
+        }
+    }
+}
 
-                this.logger.sendTelemetryEvent({
+/**
+ * This class has the following responsibilities:
```

Comment:
Thanks for this explanation! It took me a lot of code inspection to figure this out!

Reply:
not sure why this is showing up as a "review comment"